In [1]:
from dbfread import DBF
import pandas as pd
from pandas import DataFrame

## Load 10 years of accident data, from 2007 to 2016

In [2]:
dbf07= DBF('accident/accident2007.dbf')
dbf08= DBF('accident/accident2008.dbf')
dbf09= DBF('accident/accident2009.dbf')
dbf10= DBF('accident/accident2010.dbf')
dbf11 = DBF('accident/accident2011.dbf')
dbf12 = DBF('accident/accident2012.dbf')
dbf13 = DBF('accident/accident2013.dbf')
dbf14 = DBF('accident/accident2014.dbf')
dbf15 = DBF('accident/accident2015.dbf')
dbf16 = DBF('accident/accident2016.dbf')
accidents07 = DataFrame(iter(dbf07))
accidents08 = DataFrame(iter(dbf08))
accidents09 = DataFrame(iter(dbf09))
accidents10 = DataFrame(iter(dbf10))
accidents11 = DataFrame(iter(dbf11))
accidents12 = DataFrame(iter(dbf12))
accidents13 = DataFrame(iter(dbf13))
accidents14 = DataFrame(iter(dbf14))
accidents15 = DataFrame(iter(dbf15))
accidents16 = DataFrame(iter(dbf16))


First, we want to combine accidents10 ~ accidents16 to one dataframe. Since not all of the accident data downloaded from the U.S. Department of Transportation have the same features, by using the `jion:inner` option in `pd.concat` function, we can get the intersection of features.

In [4]:
# rename column name in frame07 so that columns names are the same with other frames
accidents07.rename(columns={'latitude': 'LATITUDE', 'longitud': 'LONGITUD'}, inplace=True)

In [6]:
# take a look inside how the accident data file looks like
allaccidents = pd.concat([accidents07,accidents08,accidents09,accidents10,accidents11,accidents12,accidents13,accidents14,accidents15,accidents16], axis=0,join='inner')
allaccidents.head()

STATE  COUNTY  MONTH   DAY  HOUR  MINUTE  VE_TOTAL  PERSONS  PEDS  NHS  \
0    1.0    73.0    1.0   2.0  23.0    15.0       1.0      1.0   0.0  0.0   
1    1.0    19.0    1.0  30.0  13.0     5.0       3.0      4.0   0.0  0.0   
2    1.0     9.0    1.0  14.0  12.0    13.0       2.0      5.0   0.0  0.0   
3    1.0    15.0    1.0   1.0  15.0     0.0       2.0      2.0   0.0  1.0   
4    1.0    51.0    1.0   6.0  19.0    35.0       2.0      5.0   0.0  0.0   

     ...        CITY  MILEPT    YEAR  TWAY_ID  TWAY_ID2     RAIL  VE_FORMS  \
0    ...         0.0     0.0  2007.0     1493  00000000  0000000       1.0   
1    ...       630.0     0.0  2007.0    SR-68       270  0000000       3.0   
2    ...      2330.0     0.0  2007.0  US-SR53  00000000  0000000       2.0   
3    ...         0.0  2468.0  2007.0   US-SR1      7260  0000000       2.0   
4    ...      2063.0     0.0  2007.0    SR-14      I-65  0000000       2.0   

   WEATHER   LATITUDE   LONGITUD  
0      1.0  33.455839 -87.017928  
1      1.0  34.160597 -85.678075  
2      1.0  33.975717 -86.496347  
3      1.0  33.853258 -85.915622  
4      1.0  32.488833 -86.408936  

[5 rows x 42 columns]

In [7]:
allaccidents.shape

(320874, 42)

The allaccidents table recorded 320874 accidents from 2010-2016, and it has 42 features. Here are the meaning of some of the features according to the `FARS Analytical User’s Manual`.

### Explaination of variables
*VE_TOTAL*: Number of Vehicle in crash <br/>
*VE_FORMS*: Number of Motor Vehicles in Transport (MVIT) <br/>
*PED*: Number of Persons Not in Motor Vehicles <br/>
*NHS*: National Highway System<br/>
*ROUTE*: Route Signing <br/>
*SP_JUR*: Special Jurisdiction <br/>
*HARM_EV*: First Harmful Event<br/>
*TWAY_ID , TWAY_ID2* : Trafficway Identifier <br/>
*MILEPT*: Milepoint <br/>
*SP_JUR*: Special Jurisdiction<br/>
*HARM_EV*: injury or damage producing First Harmful Event <br/> 
*MAN_COLL*:Manner of Collision  <br/> 
*RELJCT1, RELJCT2*: Relation to Junction- Within Interchange Area, Specific Location. <br/>
*TYP_INT*: Type of Intersection <br/>
*REL_ROAD*: Relation to Trafficway <br/>
*LGT_COND*: Light Condition<br/> 
*NOT_HOUR,MIN*: Min, Hour of Notification <br/>
*ARR_HOUR,MIN*: Hour, Min arrival at scene <br/>
*HOSP_HR,MIN*: Hour, Min arrival at hospital <br/>
*CF1, CF2, CF3*:Related Factors- Crash Level, factors related to the crash <br/>
*FATALS*: Fatalities<br/>
*DRUNK_DR*: Number of Drinking Drivers<br/> 
*RAIL*: Rail Grade Crossing Identifier<br/>

For more detailed information, please refer to `FARS Analytical User’s Manual`.

## Select variables and rename variables

Observed from the table above, some of the variables in the table are not very readable. Therefore, in order to make it easier to understand the variables,we renamed some of the variables according to `FARS Analytical User’s Manual`  downloaded from the  `U.S. Department of Transportation`  website.

### Load vehicle data file which contains mortality rate

We also want to study the mortality rate of fatal accidents. The data element “Fatalities in Vehicle” in the Vehicle data file from the `U.S. Department of Transportation` website provides the number of deaths in a vehicle.

In [8]:
vdbf07= DBF('vehicle_deaths/vehicle2007.dbf')
vdbf08= DBF('vehicle_deaths/vehicle2008.dbf')
vdbf09= DBF('vehicle_deaths/vehicle2009.dbf')
vdbf10= DBF('vehicle_deaths/vehicle2010.dbf')
vdbf11= DBF('vehicle_deaths/vehicle2011.dbf')
vdbf12= DBF('vehicle_deaths/vehicle2012.dbf')
vdbf13= DBF('vehicle_deaths/vehicle2013.dbf')
vdbf14= DBF('vehicle_deaths/vehicle2014.dbf')
# vdbf15= DBF('vehicle_deaths/vehicle2015.dbf')
vdbf16= DBF('vehicle_deaths/vehicle2016.dbf')
vehicle07 = DataFrame(iter(vdbf07))
vehicle08 = DataFrame(iter(vdbf08))
vehicle09 = DataFrame(iter(vdbf09))
vehicle10 = DataFrame(iter(vdbf10))
vehicle11 = DataFrame(iter(vdbf11))
vehicle12 = DataFrame(iter(vdbf12))
vehicle13 = DataFrame(iter(vdbf13))
vehicle14 = DataFrame(iter(vdbf14))
# vehicle15 = DataFrame(iter(vdbf15))
vehicle16 = DataFrame(iter(vdbf16))

In [9]:
#combine all vehicle file
allvehicle = pd.concat([vehicle07,vehicle08,vehicle09,vehicle10,vehicle11,vehicle12,vehicle13,vehicle14,vehicle16], axis=0,join='inner')
allvehicle.head(10)

VIN VIN_1 VIN_2 VIN_3 VIN_4 VIN_5 VIN_6 VIN_7 VIN_8 VIN_9  \
0  JHMCB7657NC0     J     H     M     C     B     7     6     5     7   
1  JN1HJ01P7LT4     J     N     1     H     J     0     1     P     7   
2  1G6CD53BXP42     1     G     6     C     D     5     3     B     X   
3  1FTPW12524KB     1     F     T     P     W     1     2     5     2   
4  1D4HD58266F1     1     D     4     H     D     5     8     2     6   
5  2BCCV81J7JB5     2     B     C     C     V     8     1     J     7   
6  1G2NE52T5WM5     1     G     2     N     E     5     2     T     5   
7  1FTZX17W1WNB     1     F     T     Z     X     1     7     W     1   
8  1FAFP40604F1     1     F     A     F     P     4     0     6     0   
9  1GHDX03E6XD2     1     G     H     D     X     0     3     E     6   

     ...    DR_WGT DR_HGT LAST_YR  FIRST_YR   DR_ZIP  MONTH  HARM_EV  \
0    ...     175.0   68.0  2006.0    2006.0  35117.0    1.0     42.0   
1    ...     115.0   63.0     0.0       0.0  35960.0    1.0     12.0   
2    ...     185.0   71.0     0.0       0.0  35983.0    1.0     12.0   
3    ...     195.0   72.0     0.0       0.0  35983.0    1.0     12.0   
4    ...     180.0   70.0     0.0       0.0  36203.0    1.0     12.0   
5    ...     180.0   73.0  2006.0    2004.0  35031.0    1.0     12.0   
6    ...     110.0   65.0     0.0       0.0  36271.0    1.0     12.0   
7    ...     330.0   74.0     0.0       0.0  35904.0    1.0     12.0   
8    ...     190.0   69.0  2004.0    2004.0  36067.0    1.0     12.0   
9    ...     190.0   62.0     0.0       0.0  36092.0    1.0     12.0   

   MAN_COLL  HIT_RUN  VE_FORMS  
0       0.0      0.0       1.0  
1       5.0      0.0       3.0  
2       5.0      0.0       3.0  
3       5.0      0.0       3.0  
4       7.0      1.0       2.0  
5       7.0      1.0       2.0  
6       4.0      0.0       2.0  
7       4.0      0.0       2.0  
8       5.0      0.0       2.0  
9       5.0      0.0       2.0  

[10 rows x 74 columns]

In [10]:
mortalities = allvehicle[['ST_CASE','DEATHS']]
mortalities.head()

ST_CASE  DEATHS
0  10001.0     1.0
1  10002.0     0.0
2  10002.0     2.0
3  10002.0     0.0
4  10003.0     0.0

In [11]:
mortality_cleaned = mortalities.groupby(['ST_CASE']).sum().reset_index()
mortality_cleaned.head()

ST_CASE  DEATHS
0  10001.0     9.0
1  10002.0     9.0
2  10003.0    12.0
3  10004.0     9.0
4  10005.0     8.0

In [12]:
allaccidents['DEATHS']= allaccidents.ST_CASE.map(mortality_cleaned.DEATHS)
allaccidents.head()

STATE  COUNTY  MONTH   DAY  HOUR  MINUTE  VE_TOTAL  PERSONS  PEDS  NHS  \
0    1.0    73.0    1.0   2.0  23.0    15.0       1.0      1.0   0.0  0.0   
1    1.0    19.0    1.0  30.0  13.0     5.0       3.0      4.0   0.0  0.0   
2    1.0     9.0    1.0  14.0  12.0    13.0       2.0      5.0   0.0  0.0   
3    1.0    15.0    1.0   1.0  15.0     0.0       2.0      2.0   0.0  1.0   
4    1.0    51.0    1.0   6.0  19.0    35.0       2.0      5.0   0.0  0.0   

    ...    MILEPT    YEAR  TWAY_ID  TWAY_ID2     RAIL  VE_FORMS  WEATHER  \
0   ...       0.0  2007.0     1493  00000000  0000000       1.0      1.0   
1   ...       0.0  2007.0    SR-68       270  0000000       3.0      1.0   
2   ...       0.0  2007.0  US-SR53  00000000  0000000       2.0      1.0   
3   ...    2468.0  2007.0   US-SR1      7260  0000000       2.0      1.0   
4   ...       0.0  2007.0    SR-14      I-65  0000000       2.0      1.0   

    LATITUDE   LONGITUD  DEATHS  
0  33.455839 -87.017928     4.0  
1  34.160597 -85.678075     2.0  
2  33.975717 -86.496347     2.0  
3  33.853258 -85.915622     2.0  
4  32.488833 -86.408936     3.0  

[5 rows x 43 columns]

In [13]:
allaccidents.shape

(320874, 43)